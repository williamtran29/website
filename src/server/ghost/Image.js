import { getSize } from 'server/utils/image'

class Image {
  constructor(url) {
    this.url = url
  }

  async getDimensions() {
    if (this.dimensions) return this.dimensions
    this.dimensions = getSize(this.url)
    return this.dimensions
  }

  async width() {
    return (await this.getDimensions()).width
  }

  async height() {
    return (await this.getDimensions()).height
  }
}

export default Image
