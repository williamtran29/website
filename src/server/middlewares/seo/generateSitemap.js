import sm from 'sitemap'
import { run } from 'server/graphql'
import * as routePaths from 'shared/routePaths'

const siteMapToString = sitemap =>
  new Promise((resolve, reject) => {
    sitemap.toXML((err, xml) => {
      if (err) {
        reject(err)
        return
      }

      resolve(xml.toString())
    })
  })

async function generateSitemap() {
  const data = await run(/* GraphQL */ `
    query sitemap {
      sessions {
        updatedAt
        link
      }
      articles {
        posts {
          link
          updated_at
        }
      }
    }
  `)

  return siteMapToString(
    sm.createSitemap({
      hostname: 'https://www.smooth-code.com',
      cacheTime: 600000,
      urls: [
        { url: routePaths.homeRoute(), changefreq: 'weekly', priority: 1 },
        ...data.sessions.map(session => ({
          url: session.link,
          lastmodISO: session.updatedAt,
          changefreq: 'weekly',
          priority: 0.9,
        })),
        ...data.articles.posts.map(article => ({
          url: article.link,
          lastmodISO: article.updated_at,
          changefreq: 'weekly',
          priority: 0.9,
        })),
      ],
    }),
  )
}

export default generateSitemap
