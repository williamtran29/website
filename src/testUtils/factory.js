import { factory } from 'factory-girl'
import ObjectionAdapter from 'factory-girl-objection-adapter'
import Training from 'server/models/Training'

factory.setAdapter(new ObjectionAdapter())

factory.define('training', Training, {
  rank: factory.sequence('training.rank'),
  title: factory.chance('last'),
  abstract: factory.chance('sentence'),
  description: factory.chance('paragraph'),
  objectives: factory.chance('paragraph'),
  prerequisites: factory.chance('paragraph'),
  icon: factory.chance('word'),
  slug: factory.chance('word'),
  live: true,
})

export default factory
