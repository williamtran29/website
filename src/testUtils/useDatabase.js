/* eslint-disable import/no-extraneous-dependencies */
import { truncate } from 'knex-scripts'
import Knex from 'knex'
import * as database from 'server/services/database'
import knexConfig from '../../knexfile'

function useDatabase() {
  beforeAll(async () => {
    const knex = database.connect()
    await knex.migrate.latest()
  })

  afterAll(async () => {
    await database.disconnect()
  })

  beforeEach(async () => {
    await truncate({ getKnex: () => Knex(knexConfig.test) })
  })
}

export default useDatabase
