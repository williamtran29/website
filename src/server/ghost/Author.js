/* eslint-disable camelcase */
import { trainerRoute } from 'shared/routePaths'
import Image from 'server/ghost/Image'
import Post from 'server/ghost/Post'

class Author {
  constructor(options) {
    Object.assign(this, options)

    if (this.profile_image) this.profile_image = new Image(this.profile_image)
    if (this.posts) this.posts = this.posts.map(post => new Post(post))
  }

  link() {
    return trainerRoute(this.slug)
  }
}

export default Author
