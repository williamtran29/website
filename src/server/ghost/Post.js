/* eslint-disable camelcase */
import { articleRoute } from 'shared/routePaths'
import Author from 'server/ghost/Author'
import Image from 'server/ghost/Image'

class Post {
  constructor(options) {
    Object.assign(this, options)
    if (this.author) this.author = new Author(options.author)
    if (this.feature_image)
      this.feature_image = new Image(options.feature_image)
  }

  link() {
    return articleRoute(this.slug)
  }
}

export default Post
