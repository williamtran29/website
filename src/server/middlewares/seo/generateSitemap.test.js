import useDatabase from 'testUtils/useDatabase'
import factory from 'testUtils/factory'
import * as ghostApi from 'server/ghost/ghostApi'
import generateSitemap from './generateSitemap'

jest.mock('server/ghost/ghostApi')

describe('#generateSitemap', () => {
  useDatabase()

  beforeEach(async () => {
    ghostApi.getPosts.mockImplementation(async () => ({
      posts: [
        {
          link: '/articles/test',
          updated_at: '2018-01-05T12:51:20.199Z',
        },
        {
          link: '/articles/test2',
          updated_at: '2018-06-05T12:51:20.199Z',
        },
      ],
    }))
    await factory.createMany('training', 5, {
      slug: factory.seq(),
      updated_at: factory.seq(i => `2017-01-01T0${i}:00:00.000Z`),
    })
  })

  it('should generate sitemap', async () => {
    const sitemap = await generateSitemap()
    expect(sitemap).toMatchSnapshot()
  })
})
