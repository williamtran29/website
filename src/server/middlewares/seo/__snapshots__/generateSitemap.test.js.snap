// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#generateSitemap should generate sitemap 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:mobile=\\"http://www.google.com/schemas/sitemap-mobile/1.0\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\">
<url> <loc>https://www.smooth-code.com/</loc> <changefreq>weekly</changefreq> <priority>1.0</priority> </url>
<url> <loc>https://www.smooth-code.com/articles/test</loc> <lastmod>2018-01-05T12:51:20.199Z</lastmod> <changefreq>weekly</changefreq> <priority>0.9</priority> </url>
<url> <loc>https://www.smooth-code.com/articles/test2</loc> <lastmod>2018-06-05T12:51:20.199Z</lastmod> <changefreq>weekly</changefreq> <priority>0.9</priority> </url>
</urlset>"
`;
